api/doc.go

```diff
@@ -15,72 +15,69 @@
 //
-// - Data Length is in bytes and encoded in network order.
+// • Notifications are sent by either the proxy or clients and do not generate
+// responses.
 //
-// - Reserved is reserved for future use.
+// Frame Structure
 //
-// - Data is the JSON-encoded request or response data
+// The frame format is illustrated below:
 //
-// On top of of this request/response mechanism, the proxy defines payloads,
-// which are effectively the various function calls defined in the API.
+//                      1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+//  0 1 2 3 4 5 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  ┌───────────────────────────┬───────────────┬───────────────┐
+//  │          Version          │ Header Length │   Reserved    │
+//  ├───────────────────────────┼─────┬─┬───────┼───────────────┤
+//  │          Reserved         │ Res.│E│ Type  │    Opcode     │
+//  ├───────────────────────────┴─────┴─┴───────┴───────────────┤
+//  │                      Payload Length                       │
+//  ├───────────────────────────────────────────────────────────┤
+//  │                                                           │
+//  │                         Payload                           │
+//  │                                                           │
+//  │      (variable length, optional and opcode-specific)      │
+//  │                                                           │
+//  └───────────────────────────────────────────────────────────┘
 //
-// Requests have 2 fields: the payload id (function name) and its data
-// (function argument(s))
+// All header fields are encoded in network order (big endian).
 //
-//  type Request struct {
-//      Id    string          `json:"id"`
-//      Data *json.RawMessage `json:"data,omitempty"`
-//  }
+// • Version (16 bits) is the proxy protocol version. See api.Version for
+// details about what information it encodes.
 //
-// Responses have 3 fields: success, error and data.
+// • Header Length (8 bits) is the length of the header in number of 32 bits
+// words.  Header Length is greater or equal to 3 (12 bytes).
 //
-//  type Response struct {
-//      Success bool                   `json:"success"`
-//      Error   string                 `json:"error,omitempty"`
-//      Data    map[string]interface{} `json:"data,omitempty"`
-//  }
+// • Type (4 bits) is the frame type: command (0x0), response (0x1),
+// stream (0x2) or notification (0x3).
 //
-// Unsurprisingly, the response has the result of a command, with success
-// indicating if the request has succeeded for not. If success is true, the
-// response can carry additional return values in data. If success if false,
-// error will contain an error string suitable for reporting the error to a
-// user.
+// • Opcode (8 bits) specifies the kind of command, response, stream or
+// notification this frame represents. In conjunction with Type, this field
+// will dictate the payload content.
```

Comment:
@dlespiau Is it safe to assume that the response will have the same opcode as the Command for which the response is being sent? I am asking this to handle the response from ConnectShim command, the shim is not waiting for the response to be sent before processing other frames from the proxy like we had discussed.
But in case an error is sent later for ConnectShim, the shim does need to terminate.